Userland/Libraries/LibJS/Runtime/Object.cpp

```diff
@@ -113,881 +68,1009 @@ Object::~Object()
+bool Object::is_extensible() const
 {
-    return shape().prototype();
+    return internal_is_extensible();
 }
 
-const Object* Object::prototype() const
+// 7.3 Operations on Objects, https://tc39.es/ecma262/#sec-operations-on-objects
+
+// 7.3.2 Get ( O, P ), https://tc39.es/ecma262/#sec-get-o-p
+Value Object::get(PropertyName const& property_name) const
 {
-    return shape().prototype();
+    VERIFY(property_name.is_valid());
+
+    // 1. Assert: Type(O) is Object.
+    // 2. Assert: IsPropertyKey(P) is true.
+
+    // 3. Return ? O.[[Get]](P, O).
+    return internal_get(property_name, this);
 }
 
-// 10.1.2.1 OrdinarySetPrototypeOf ( O, V ), https://tc39.es/ecma262/#sec-ordinarysetprototypeof
-bool Object::set_prototype(Object* new_prototype)
+// 7.3.3 GetV ( V, P ) is defined as Value::get().
+
+// 7.3.4 Set ( O, P, V, Throw ), https://tc39.es/ecma262/#sec-set-o-p-v-throw
+bool Object::set(PropertyName const& property_name, Value value, bool throw_exceptions)
 {
-    if (prototype() == new_prototype)
-        return true;
-    if (!m_is_extensible)
-        return false;
-    auto* prototype = new_prototype;
-    while (prototype) {
-        if (prototype == this)
-            return false;
-        // NOTE: This is a best-effort implementation of the following step:
-        // "If p.[[GetPrototypeOf]] is not the ordinary object internal method defined in 10.1.1,
-        // set done to true."
-        // We don't have a good way of detecting whether certain virtual Object methods have been
-        // overridden by a given object, but as ProxyObject is the only one doing that, this check
-        // does the trick.
-        if (is<ProxyObject>(prototype))
-            break;
-        prototype = prototype->prototype();
+    VERIFY(property_name.is_valid());
+    VERIFY(!value.is_empty());
+    auto& vm = this->vm();
+
+    // 1. Assert: Type(O) is Object.
+    // 2. Assert: IsPropertyKey(P) is true.
+    // 3. Assert: Type(Throw) is Boolean.
+
+    // 4. Let success be ? O.[[Set]](P, V, O).
+    auto success = internal_set(property_name, value, this);
+    if (vm.exception())
+        return {};
+
+    // 5. If success is false and Throw is true, throw a TypeError exception.
+    if (!success && throw_exceptions) {
+        // FIXME: Improve/contextualize error message
+        vm.throw_exception<TypeError>(global_object(), ErrorType::ObjectSetReturnedFalse);
+        return {};
     }
-    auto& shape = this->shape();
-    if (shape.is_unique())
-        shape.set_prototype_without_transition(new_prototype);
-    else
-        m_shape = shape.create_prototype_transition(new_prototype);
-    return true;
+
+    // 6. Return success.
+    return success;
 }
 
-bool Object::has_prototype(const Object* prototype) const
+// 7.3.5 CreateDataProperty ( O, P, V ), https://tc39.es/ecma262/#sec-createdataproperty
+bool Object::create_data_property(PropertyName const& property_name, Value value)
```

Comment:
This seems to be the only method without a `VERIFY(property_name.is_valid());` at the beginning

Reply:
Fixed, thanks!